statistics/cmsketch.go

```diff
@@ -44,6 +58,81 @@ func NewCMSketch(d, w int32) *CMSketch {
 	return &CMSketch{depth: d, width: w, table: tbl}
 }
 
+// NewCMSketchWithTopN returns a new CM sketch with TopN elements.
+// Only when the number of topn elements exceeded len(data) / topnThreshold will store topn elements
+func NewCMSketchWithTopN(d, w int32, data [][]byte, n uint32) *CMSketch {
+	tbl := make([][]uint32, d)
+	for i := range tbl {
+		tbl[i] = make([]uint32, w)
+	}
+	c := &CMSketch{depth: d, width: w, table: tbl}
+	c.BuildTopN(data, n, TopNThreshold)
+	return c
+}
+
+// BuildTopN builds table of top N elements.
+// elements in data should not be modified after this call.
+// See https://pdfs.semanticscholar.org/5ae2/6358989c6595798da8f8fd6c2a517536bcae.pdf
+func (c *CMSketch) BuildTopN(data [][]byte, n uint32, topnThreshold uint32) {
+	c.topn = make([]cmscount, n)
+	for k := range data {
+		found, pos := false, -1
+		for i := range c.topn {
+			if bytes.Equal(c.topn[i].data, data[k]) {
+				found = true
+				pos = i
+				break
+			} else if c.topn[i].count == 0 {
+				pos = i
+			}
+		}
+		if found {
+			c.topn[pos].count++
+		} else if !found && pos != -1 {
+			c.topn[pos] = cmscount{data[k], 1}
+		} else {
+			for i := range c.topn {
+				c.topn[i].count--
+			}
+		}
+	}
+
+	for i := range c.topn {
+		c.topn[i].count = 0
+	}
+
+	topncount := uint32(0)
+
+	for k := range data {
+		found := false
+		for i := range c.topn {
+			if bytes.Equal(c.topn[i].data, data[k]) {
+				found = true
+				c.topn[i].count++
+				c.count++
+				topncount++
+				break
+			}
+		}
+		if !found {
+			c.InsertBytes(data[k])
+		}
+	}
+
+	if n/topnThreshold > topncount {
+		for i := range c.topn {
+			c.InsertBytesN(c.topn[i].data, c.topn[i].count)
+		}
+	} else {
+		c.topnindex = make(map[hack.MutableString]uint32)
+		for i := range c.topn {
+			c.topnindex[hack.String(c.topn[i].data)] = c.topn[i].count
+		}
+	}
+
+	c.topn = nil
+}
+
 // InsertBytes inserts the bytes value into the CM Sketch.
 func (c *CMSketch) InsertBytes(bytes []byte) {
```

Comment:
How about let `InsertBytes` call `InsertBytesN(bytes, 1)` directly, which is better for further maintaining.